### PR TITLE
Ensure webhook service routing resolves kubernetes.default.svc correctly

### PR DESCRIPTION
**Pull request by liggitt**
Going through the normal endpoint resolve path isn't correct in multi-master scenarios

The auth wrapper is pulling from LoopbackClientConfig, the service resolver should do the same

```release-note
Fixes the kubernetes.default.svc loopback service resolution to use a loopback configuration.
```

### CODE REVIEW

**Comment by lavalamp** [comment]
Please test these things if you want them to keep working:

/lgtm cancel

I think this is an awful hack to work around not having an identity for apiserver. e.g. clearly this can't work for an aggregated apiserver. Or am I misunderstanding something?

**Comment by liggitt** [comment]
> I think this is an awful hack to work around not having an identity for apiserver

the only service the apiserver can make authentication assumptions about is the kubernetes/default service, and only because it is serving it and controls authentication for it.

> e.g. clearly this can't work for an aggregated apiserver

that's why this is only used in kube-apiserver. aggregated api servers running on top of the kube control plane use default dns resolution and in-cluster auth without incident.

**Comment by lavalamp** [comment]
> the only service the apiserver can make authentication assumptions about is the kubernetes/default service, and only because it is serving it and controls authentication for it.

A concrete example request that exercises the path would help a lot.

**Comment by liggitt** [comment]
test added, three additional bugs fixed. most integration tests do not actually use the real apiserver construction methods, so wouldn't have exercised the code that was fixed. 

The closest integration test to reality I found was this one:

https://github.com/kubernetes/kubernetes/blob/bf3cda66c3664dd874c8f2e41a99cbab96e74b89/test/integration/examples/apiserver_test.go#L99-L140

In the test commit in this PR, I started to extract the test setup to something that:
1. gives you a one-call invocation to get a running apiserver
2. lets you tweak run options or config prior to starting
3. makes use of the actual config and server-building functions that kube-apiserver uses

https://github.com/kubernetes/kubernetes/blob/596e4c6c48316a808ff3c16afdf4c756d572e204/test/integration/examples/webhook_test.go#L43-L62

File: staging/src/k8s.io/client-go/transport/transport.go [hunk 1]
```diff
@@ -52,7 +52,7 @@ func New(config *Config) (http.RoundTripper, error) {
 // TLSConfigFor returns a tls.Config that will provide the transport level security defined
 // by the provided Config. Will return nil if no transport level security is requested.
 func TLSConfigFor(c *Config) (*tls.Config, error) {
-	if !(c.HasCA() || c.HasCertAuth() || c.TLS.Insecure) {
+	if !(c.HasCA() || c.HasCertAuth() || c.TLS.Insecure || len(c.TLS.ServerName) > 0) {
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/transport/transport.go]: if tls.Insecure is true, why don't we just return nil, nil?

**Comment by deads2k** [comment]
Question on the TLS bit.  Integration test looks like it is testing the wiring more than the code since kube integration doesn't run real servers, but sure.

File: staging/src/k8s.io/client-go/transport/transport.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/transport/transport.go]: anything that deviates from the defaults must return a custom tls.Config

setting Insecure=true or a custom ServerName requires a non-default tls.Config

**Comment by liggitt** [comment]
any other comments?

**Comment by liggitt** [comment]
rebased

**Comment by munnerz** [comment]
Is this likely to be also accepted into release-1.9 if cherry picked? Not sure what the usual convention is for which releases something should be backported to.

**Comment by liggitt** [comment]
> Is this likely to be also accepted into release-1.9 if cherry picked? Not sure what the usual convention is for which releases something should be backported to.

picked to 1.9 in https://github.com/kubernetes/kubernetes/pull/63219

**Pull request merged**